### ISSUE TITLE
Only report coverage for TS files that have spec files?

### ISSUE THREAD

**Issue by jgodi** [monounity/karma-typescript#116]
I have a few files in my library that don't need/have spec files and they are showing in my coverage report. How can I ignore these without specifically calling them out with the `exclude`?

```
-----------------------|----------|----------|----------|----------|----------------|
File                   |  % Stmts | % Branch |  % Funcs |  % Lines |Uncovered Lines |
-----------------------|----------|----------|----------|----------|----------------|
 src/                  |    10.53 |      100 |        0 |    13.33 |                |
  index.ts             |        0 |      100 |      100 |        0 |        2,3,6,7 |
  ng-starter.module.ts |    15.38 |      100 |        0 |    18.18 |... 19,21,22,24 |
 src/meow/             |    33.33 |      100 |    33.33 |    35.71 |                |
  index.ts             |        0 |      100 |      100 |        0 |            1,2 |
  meow.component.ts    |      100 |      100 |      100 |      100 |                |
  meow.module.ts       |    11.11 |      100 |        0 |     12.5 |... ,7,14,15,17 |
-----------------------|----------|----------|----------|----------|----------------|
All files              |    21.62 |      100 |    16.67 |    24.14 |                |
-----------------------|----------|----------|----------|----------|----------------|
```

I only want the `meow.component.ts` to appear.

**Comment by erikbarke** [comment]
Hey @jgodi, the only way to do it (currently) is excluding files using  `karmaTypescriptConfig.coverageOptions.exclude`. Are you looking for at way to automatically exclude files that don't have a corresponding spec file?

**Comment by jgodi** [comment]
Yes, that would be amazing. I can hard-code the exclusion, but it would be nice to only include coverage for files that have tests with them.

Especially in Angular, there are module files, index files, etc that only have decorators on a class. Testing these would provide no benefit, but including them reduces the numbers drastically.

**Comment by erikbarke** [comment]
Ok, so something like

```js
karmaTypescriptConfig {
    coverageOptions: {
        exclude: [/\.(d|module|spec)\.ts$/i, /index\.ts$/i]
    }
}
```

to exclude all `*.module.ts` and `index.ts` files wouldn't be enough to solve your issue?

If so, this would have to be an opt-in feature because it would mean that you could have 100% test coverage even with a ton of untested files, if all untested files are excluded automatically.

These are good starting points for a PR:

- [src/karma/preprocessor.ts](https://github.com/monounity/karma-typescript/blob/next/src/karma/preprocessor.ts) - this is where the compiler, the bundler and the coverage instrumentation are tied together
- [src/bundler/dependency-walker.ts](https://github.com/monounity/karma-typescript/blob/next/src/bundler/dependency-walker.ts) - this is where all dependencies between all Typescript files are collected
- [src/istanbul/coverage.ts](https://github.com/monounity/karma-typescript/blob/next/src/istanbul/coverage.ts) - this is where files are instrumented or excluded from coverage

Cheers!

**Comment by teogeos** [comment]
Just so you know, I prefer coverage to show for all files by default. I think it's better to make it harder to *forget* to test files by default. I'd rather have to explicitly exclude files.
